oras/auth/base.py

```diff
@@ -2,37 +2,49 @@
 __copyright__ = "Copyright The ORAS Authors."
 __license__ = "Apache-2.0"
 
-
-from typing import Optional
+import abc
+from typing import Dict, Optional, Tuple
 
 import oras.auth.utils as auth_utils
 import oras.container
-import oras.decorator as decorator
 from oras.logger import logger
-from oras.types import container_type
+import oras.utils
+
+import requests
 
 
-class AuthBackend:
+class AuthBackend(abc.ABC):
     """
     Generic (and default) auth backend.
     """
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, session: requests.Session):
         self._auths: dict = {}
+        self.session = session
+        self.headers: Dict[str, str] = {}
+
+    @abc.abstractmethod
+    def authenticate_request(
+        self,
+        original: requests.Response,
+        headers: Optional[dict[str, str]] = None,
+        refresh=False,
+    ) -> Tuple[dict[str, str], bool]:
+        pass
 
-    def get_auth_header(self):
-        raise NotImplementedError
-
-    def get_container(self, name: container_type) -> oras.container.Container:
+    def set_header(self, name: str, value: str):
```

Comment:
This method was referenced in multiple places but did not actually exist. So I added it by copying it from Provider.